clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java

```diff
@@ -0,0 +1,166 @@
+   */
+  Set<String> getClusters(AmazonECS ecs, ProviderCache providerCache) {
+    Set<String> clusters = providerCache.getAll(ECS_CLUSTERS.toString()).stream()
+      .map(cacheData -> (String) cacheData.getAttributes().get("clusterArn"))
+      .collect(Collectors.toSet());
+
+    if (clusters == null || clusters.isEmpty()) {
+      clusters = new HashSet<>();
+      String nextToken = null;
+      do {
+        ListClustersRequest listClustersRequest = new ListClustersRequest();
+        if (nextToken != null) {
+          listClustersRequest.setNextToken(nextToken);
+        }
+        ListClustersResult listClustersResult = ecs.listClusters(listClustersRequest);
+        clusters.addAll(listClustersResult.getClusterArns());
+
+        nextToken = listClustersResult.getNextToken();
+      } while (nextToken != null && nextToken.length() != 0);
+    }
+
+    return clusters;
+  }
+
+  /**
+   * Provides the key namespace that the caching agent is authoritative of.
+   * Currently only supports the caching agent being authoritative over one key namespace.
+   * @return Key namespace.
+   */
+  String getAuthoritativeKeyName() {
+    Collection<AgentDataType> authoritativeNamespaces = getProvidedDataTypes().stream()
+      .filter(agentDataType -> agentDataType.getAuthority().equals(AUTHORITATIVE))
+      .collect(Collectors.toSet());
+
+    if (authoritativeNamespaces.size() != 1) {
+      throw new RuntimeException("AbstractEcsCachingAgent supports only one authoritative key namespace.");
+    }
+
+    return authoritativeNamespaces.iterator().next().getTypeName();
+  }
+
+  CacheResult buildCacheResult(String authoritativeKeyName, List<T> items, ProviderCache providerCache) {
+    String prettyKeyName = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, authoritativeKeyName);
+
+    Map<String, Collection<CacheData>> dataMap = generateFreshData(items);
+
+    Set<String> oldKeys = providerCache.getAll(authoritativeKeyName).stream()
+      .map(cache -> cache.getId()).collect(Collectors.toSet());
+
+    Map<String, Collection<String>> evictions = computeEvictableData(dataMap.get(authoritativeKeyName), oldKeys);
+    log.info("Evicting " + evictions.size() + " " + prettyKeyName + (evictions.size() > 1 ? "s" : "") + " in " + getAgentType());
+
+    return new DefaultCacheResult(dataMap, evictions);
+  }
+
+  /**
+   * Evicts cache that does not belong to an entity on the ECS service.
+   * This is done by evicting old keys that are no longer found in the new keys provided by the new data.
+   * @param newData New data that contains new keys.
+   * @param oldKeys Old keys.
+   * @return Key collection associated to the key namespace the the caching agent is authoritative of.
+   */
+  private Map<String, Collection<String>> computeEvictableData(Collection<CacheData> newData, Collection<String> oldKeys) {
+    Set<String> newKeys = newData.stream().map(newKey -> newKey.getId()).collect(Collectors.toSet());
```

Comment:
`newData.stream().map(CacheData::getId).collect(Collectors.toSet())` is a bit cleaner.

Reply:
Done.